src/lfortran/semantics/ast_body_visitor.cpp

```diff
@@ -1075,7 +1075,44 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         tmp = (ASR::asr_t*)ASRUtils::STMT(ASR::make_Assignment_t(al, x.base.base.loc, target_var, LFortran::ASRUtils::EXPR(ASR::make_IntegerConstant_t(al, x.base.base.loc, x.m_assign_label, int32_type)), nullptr));
     }
 
+    bool is_statement_function( const AST::Assignment_t &x ) {
+        if (!AST::is_a<AST::FuncCallOrArray_t>(*x.m_target)) {
+            return false;
+        } else {
+            // Look for the type of *x.m_target in symbol table, if it is integer or nullptr then it is a statement function
+            std::string var_name = "dfloat";
+
+            ASR::symbol_t *sym = current_scope->resolve_symbol(var_name);
+            if (sym==nullptr) {
+                if (compiler_options.implicit_typing) {
+                    return true;
+                } else {
+                    return false;
+                }
+            } else {
+                if (ASR::is_a<ASR::Variable_t>(*sym)) {
+                    auto v = ASR::down_cast<ASR::Variable_t>(sym);
+                    if (ASR::is_a<ASR::Integer_t>(*v->m_type) or ASR::is_a<ASR::Real_t>(*v->m_type)) {
+                        return true;
```

Comment:
In here you have to check dimensions:
```suggestion
                    if (ASRUtils::extract_dimensions_from_ttype(v->m_type, dims) == 0) {
                        return true;
                    } else {
                        return false;
                    }
```

Reply:
Actually, just do `if (ASRUtils::is_array(v->m_type))`.